libraries/botbuilder-ai/src/luisRecognizer.ts

```diff
@@ -119,12 +120,37 @@ export class LuisRecognizer {
 
     /**
      * Creates a new LuisRecognizer instance.
-     * @param application An object conforming to the [LuisApplication](#luisapplication) definition.
+     * @param application An object conforming to the [LuisApplication](#luisapplication) definition or a string representing a LUIS application endpoint, usually retrieved from https://luis.ai.
      * @param options (Optional) options object used to control predictions. Should conform to the [LuisPrectionOptions](#luispredictionoptions) definition.
      * @param includeApiResults (Optional) flag that if set to `true` will force the inclusion of LUIS Api call in results returned by [recognize()](#recognize). Defaults to a value of `false`.
      */
-    constructor(application: LuisApplication, options?: LuisPredictionOptions, includeApiResults?: boolean) {
-        this.application = application;
+    constructor(application: string, options?: LuisPredictionOptions, includeApiResults?: boolean);
+    constructor(application: LuisApplication, options?: LuisPredictionOptions, includeApiResults?: boolean);
```

Comment:
This class allows writes to `this.application` but does not update `this.luisClient` which can lead to unpredictable behavior.
```js
const luisRecognizer = new LuisRecognizer({applicationId: "234", endpoint:"http", endpointKey:"234"}); 
luisRecognizer.application = {}; // oops! no validation, no regeneration of the luisClient. This is writable but should not be.
```
A solution would be to create accessors for the `application` property that performs validation and generates the `luisClient` when set then `Object.freeze()` the passed in `application` to make it immutable.
```js
private _application: LuisApplication;
public set application(value: LuisApplication | string) {
  this._luisApplication = this.validateAndParseLuisApplication(value);
  // validate, create a luisClient then Object.freeze(this._application) to make it immutable 
}
public get application(): LuisApplication {
 return this._luisApplication;
}
```

Reply:
can't we just make this.application setter private? if you want to change it create a new LuisRecognizer

Reply:
The visibility of this field is `private` but that does not prevent someone from making changes to a passed in Object.
```js
const config = {applicationId: "234", endpoint:"http", endpointKey:"234"};
const luisRecognizer = new LuisRecognizer(config); 
config.endpoint = ""; // <-- I can change it here
(luisRegognizer as any).application.endpoint = ""; // <--- I can change it here too
```
`Object.freeze()` makes this object immutable and throws at runtime if someone tries.

Reply:
Per offline discussions with @justinwilaby, instead extracted the values from a passed in LuisApplication instance and then constructed the LuisRecognizer.application instance with primitive values to avoid this problem.